### ISSUE TITLE
Timing issue when reading the STM32L431 flash size gets result 0

### ISSUE THREAD

**Issue by bmarvo** [stlink-org/stlink#801]
- [X] Programmer/board type: Stlink/v2
- [X] Programmer firmware version: V2J29S7 (what st-link utility says)
- [X] Operating system: Windows 10 Pro 
- [X] Stlink tools version and/or git commit hash: git-cf67780
- [X] Stlink commandline tool name: `st-info`
- [X] Target chip (and optional board): STM32L431CC (custom board)

Flash size for STM32L431 reads 0. 

Output:

```
C:\Development\stlink\scripts>st-info.exe --flash
0x0
```

I did some digging and found that I was able to get the flash size to read correctly every time by adding a delay before the read in common.c :

```c
// These are fixed...
sl->flash_base = STM32_FLASH_BASE;
sl->sram_base = STM32_SRAM_BASE;
delay(10);
stlink_read_debug32(sl,(params->flash_size_reg) & ~3, &flash_size);
if (params->flash_size_reg & 2)
    flash_size = flash_size >>16;
flash_size = flash_size & 0xffff;
```

With the delay:

```
C:\Development\stlink\scripts>st-info.exe --flash
0x40000
```

The delay is 10 milliseconds, but I just arbitrarily chose it.

I should also note that all other parameters (sram size, chip id, etc.) seem to read just fine as is.

**Comment by xor-gate** [comment]
This is really weird, by changing timing it works.

**Comment by xor-gate** [comment]
What is your libusb version, and which platform are you using e.g mingw, cygwin, visual studio?

**Comment by bmarvo** [comment]
libusb version 1.0.22

I have tested with MinGW-w64 and Visual Studio 2017. Same issue with both. Same fix works with both as well.

**Comment by bmarvo** [comment]
I can confirm that the same issue exists when building with GCC 8.3.0 on Arch Linux. The same fix also works.

**Comment by xor-gate** [comment]
Thanks @bmarvo for investigating this more, good it has the same behaviour on two independent systems and configurations. Have you tried updating the firmware of the programmer to the latest? It is also common programmers itself can have bugs programmed too :=D

**Comment by bmarvo** [comment]
@xor-gate I updated my programmers to the latest firmware (which stsw-link007 says is V2.J34.S7) and I'm still getting 0 for flash size without the delay.

**Comment by YcSmile** [comment]
i had the same question.And i build  the st-flash with a version of v1.5.1-28-gd040db5 in ubuntu 16.04 and gcc 5.4.0 20160609. 
when read the stm32f103 chip flash size reg , delay >5 ms or do a stlink_reset solved my problem.

**Comment by Nightwalker-87** [comment]
@bmarvo Can you verify this with Release v1.6.0?

**Comment by Nightwalker-87** [comment]
@bmarvo Can you verify this with Release v1.6.0?
Please also check if there is a correlation to issue 290.

**Comment by Nightwalker-87** [comment]
@YcSmile: I'll check the state from the commit you cited previously. It is likely that this parameter has been changed several times since then. May I ask you to test some later versions, as soon as we gain some idea on more recent changes in the code related to this pre-set delay?

**Comment by Nightwalker-87** [comment]
Two questions from my side:
- Can this behaviour be reproduced with other chips or boards?
- Is it really related to `libusb`?

**Comment by Ant-ON** [comment]
There is some strange behavior. Most likely it is related with a reset and may be fixed by #1071

**Comment by Nightwalker-87** [comment]
Assumed to be closed by #1071.
